ggml-metal.m

```diff
@@ -994,61 +839,129 @@ void ggml_metal_graph_compute(
     }
 }
 
-bool ggml_backend_metal_map_buffer(
-        struct ggml_backend * backend,
-        const char * name,
-        void * data,
-        size_t   size,
-        size_t   max_size) {
-    return ggml_metal_add_buffer(backend->context, name, data, size, max_size);
-}
-
 static const char * ggml_backend_metal_name(struct ggml_backend * ctx) {
     return "Metal";
 
     UNUSED(ctx);
 }
 
+static void ggml_backend_metal_free(struct ggml_backend * backend) {
+    struct ggml_metal_context * ctx_metal = (struct ggml_metal_context *)backend->context;
+    ggml_metal_free(ctx_metal);
+    free(backend);
+}
+
+static const size_t TENSOR_ALIGNMENT = 128;
+
+struct ggml_metal_buffer_wrapper {
+    id<MTLBuffer> buffer;
+};
+
+static void ggml_backend_metal_free_buffer(struct ggml_backend_buffer * alloc) {
+    struct ggml_metal_buffer_wrapper * wrapper = (struct ggml_metal_buffer_wrapper *)alloc->backend_data;
+    [wrapper->buffer release];
+    free(wrapper);
+}
+
+static struct ggml_backend_buffer * ggml_backend_metal_alloc_buffer(struct ggml_backend * backend, size_t size) {
+    struct ggml_metal_context * ctx_metal = (struct ggml_metal_context *)backend->context;
+
+    struct ggml_metal_buffer_wrapper * wrapper = malloc(sizeof(struct ggml_metal_buffer_wrapper));
+    wrapper->buffer = [ctx_metal->device newBufferWithLength:size options:MTLResourceStorageModeShared];
+
+    printf("XXXXXXXXXXXXXXX ALOC: %p %p\n", (void * )wrapper, (void *)&wrapper->buffer);
+
+    struct ggml_backend_buffer * buffer = ggml_allocator_simple_init(wrapper, size, TENSOR_ALIGNMENT);
```

Comment:
I think passing `wrapper` here may not be right. It is supposed to a pointer to the start of the buffer. Looking [here](https://developer.apple.com/documentation/metal/mtlbuffer) it may be the property `gpuAddress`.
It's fine if it is a device-specific pointer not accessible from the CPU, the allocator never dereferences it.